autoendpoint/src/routers/apns/router.rs

```diff
@@ -0,0 +1,536 @@
+                } else {
+                    warn!("APNS error: {:?}", response.error);
+                    self.incr_error_metric("response_error", channel);
+                }
+            }
+            a2::Error::ConnectionError => {
+                error!("APNS connection error");
+                self.incr_error_metric("connection", channel);
+            }
+            _ => {
+                warn!("Unknown error while sending APNS request: {}", error);
+                self.incr_error_metric("unknown", channel);
+            }
+        }
+
+        ApiError::from(ApnsError::ApnsUpstream(error))
+    }
+}
+
+#[async_trait(?Send)]
+impl Router for ApnsRouter {
+    fn register(
+        &self,
+        router_input: &RouterDataInput,
+        app_id: &str,
+    ) -> Result<HashMap<String, Value>, RouterError> {
+        if !self.clients.contains_key(app_id) {
+            return Err(ApnsError::InvalidReleaseChannel.into());
+        }
+
+        let mut router_data = HashMap::new();
+        router_data.insert(
+            "token".to_string(),
+            serde_json::to_value(&router_input.token).unwrap(),
+        );
+        router_data.insert(
+            "rel_channel".to_string(),
+            serde_json::to_value(app_id).unwrap(),
+        );
+
+        if let Some(aps) = &router_input.aps {
+            if APS::deserialize(aps).is_err() {
+                return Err(ApnsError::InvalidApsData.into());
+            }
+
+            router_data.insert("aps".to_string(), aps.clone());
+        }
+
+        Ok(router_data)
+    }
+
+    async fn route_notification(&self, notification: &Notification) -> ApiResult<RouterResponse> {
+        debug!(
+            "Sending APNS notification to UAID {}",
+            notification.subscription.user.uaid
+        );
+        trace!("Notification = {:?}", notification);
+
+        // Build message data
+        let router_data = notification
+            .subscription
+            .user
+            .router_data
+            .as_ref()
+            .ok_or(ApnsError::NoDeviceToken)?;
+        let token = router_data
+            .get("token")
+            .and_then(Value::as_str)
+            .ok_or(ApnsError::NoDeviceToken)?;
+        let channel = router_data
+            .get("rel_channel")
+            .and_then(Value::as_str)
+            .ok_or(ApnsError::NoReleaseChannel)?;
+        let aps: APS<'_> = router_data
+            .get("aps")
+            .and_then(|value| APS::deserialize(value).ok())
+            .unwrap_or_else(Self::default_aps);
+        let mut message_data = build_message_data(notification, self.settings.max_data)?;
+        message_data.insert("ver", notification.message_id.clone());
+
+        // Get client and build payload
+        let ApnsClientData { client, topic } = self
+            .clients
+            .get(channel)
+            .ok_or(ApnsError::InvalidReleaseChannel)?;
+        let payload = Payload {
```

Comment:
I know that we're checking subscription data size with the app, but each bridge also has a data limit which is smaller than the general message size (this is because we may need to re-encode the message being sent through, which can reduce the effective total message size.)

I don't see where we're checking against that, except if the service rejects the message, which could result in user confusion. Am I missing something?

Reply:
No, you're right we should be checking that here.

Reply:
It looks like the Python code does not do any size checks for APNS or ADM, and only checks the notification `data` payload in the FCM router:
https://github.com/mozilla-services/autopush/blob/91db1fa6c564af84d7dbbdb4e20e0619579edf0b/autopush/router/fcm_v1.py#L85

I'll fix this issue for the APNS router and make an issue to fix the FCM router (Rust). Should we also go back and fix the Python routers?

Reply:
#203

Reply:
Probably not a bad idea to fix up the python router too. https://github.com/mozilla-services/autopush/issues/1422